feat(platform): Add Twitter integration

Abhi1992002 opened the pull request:
- Resolves #8326  

Create a Twitter integration with some small frontend changes.

### Changes
1. Add Twitter OAuth 2.0 with PKCE support for authentication.
2. Add a way to multi-select from a list of enums by creating a multi-select on the frontend.
3. Add blocks for Twitter integration.
4. `_types.py` for repetitive enums and input types.
5. `_builders.py` for creating parameters without repeating the same logic.
6. `_serializer.py` to serialize the Tweepy enums into dictionaries so they can travel easily from Pyro5.
7. `_mappers.py` to map the frontend values to the correct request values.

> I have added a new multi-select feature because my list contains many items, and selecting all of them makes the block cluttered. This new block displays only the first two items and then show something like "2 more" . It works only for list of enums.


### Blocks

Block Name | What It Does | Error Reason | Manual Testing
-- | -- | -- | --
`TwitterBookmarkTweetBlock` | Bookmark a tweet on Twitter | No error | ✅
`TwitterGetBookmarkedTweetsBlock` | Get all your bookmarked tweets from Twitter | No error | ✅
`TwitterRemoveBookmarkTweetBlock` | Remove a bookmark for a tweet on Twitter | No error | ✅
`TwitterHideReplyBlock` | Hides a reply of one of your tweets | No error | ✅
`TwitterUnhideReplyBlock` | Unhides a reply to a tweet | No error | ✅
`TwitterLikeTweetBlock` | Likes a tweet | No error | ✅
`TwitterGetLikingUsersBlock` | Gets information about users who liked one of your tweets | No error | ✅
`TwitterGetLikedTweetsBlock` | Gets information about tweets liked by you | No error | ✅
`TwitterUnlikeTweetBlock` | Unlikes a tweet that was previously liked | No error | ✅
`TwitterPostTweetBlock` | Create a tweet on Twitter with the option to include one additional element such as media, quote, or deep link. | No error | ✅
`TwitterDeleteTweetBlock` | Deletes a tweet on Twitter using Twitter ID | No error | ✅
`TwitterSearchRecentTweetsBlock` | Searches all public Tweets in Twitter history | No error | ✅
`TwitterGetQuoteTweetsBlock` | Gets quote tweets for a specified tweet ID | No error | ✅
`TwitterRetweetBlock` | Retweets a tweet on Twitter | No error | ✅
`TwitterRemoveRetweetBlock` | Removes a retweet on Twitter | No error | ✅
`TwitterGetRetweetersBlock` | Gets information about who has retweeted a tweet | No error | ✅
`TwitterGetUserMentionsBlock` | Returns Tweets where a single user is mentioned, just put that user ID | No error | ✅
`TwitterGetHomeTimelineBlock` | Returns a collection of the most recent Tweets and Retweets posted by you and users you follow | No error | ✅
`TwitterGetUserTweetsBlock` | Returns Tweets composed by a single user, specified by the requested user ID | No error | ✅
`TwitterGetTweetBlock` | Returns information about a single Tweet specified by the requested ID | No error | ✅
`TwitterGetTweetsBlock` | Returns information about multiple Tweets specified by the requested IDs | No error | ✅
`TwitterUnblockUserBlock` | Unblock a specific user on Twitter | No error | ✅
`TwitterGetBlockedUsersBlock` | Get a list of users who are blocked by the authenticating user | No error | ✅
`TwitterBlockUserBlock` | Block a specific user on Twitter | No error | ✅
`TwitterUnfollowUserBlock` | Allows a user to unfollow another user specified by target user ID | No error | ✅
`TwitterFollowUserBlock` | Allows a user to follow another user specified by target user ID | No error | ✅
`TwitterGetFollowersBlock` | Retrieves a list of followers for a specified Twitter user ID | Need Enterprise level access | ❌
`TwitterGetFollowingBlock` | Retrieves a list of users that a specified Twitter user ID is following | Need Enterprise level access | ❌
`TwitterUnmuteUserBlock` | Allows a user to unmute another user specified by target user ID | No error | ✅
`TwitterGetMutedUsersBlock` | Returns a list of users who are muted by the authenticating user | No error | ✅
`TwitterMuteUserBlock` | Allows a user to mute another user specified by target user ID | No error | ✅
`TwitterGetUserBlock` | Gets information about a single Twitter user specified by ID or username | No error | ✅
`TwitterGetUsersBlock` | Gets information about multiple Twitter users specified by IDs or usernames | No error | ✅
`TwitterSearchSpacesBlock` | Returns live or scheduled Spaces matching specified search terms [for a week only] | No error | ✅
`TwitterGetSpacesBlock` | Gets information about multiple Twitter Spaces specified by Space IDs or creator user IDs | No error | ✅
`TwitterGetSpaceByIdBlock` | Gets information about a single Twitter Space specified by Space ID | No error | ✅
`TwitterGetSpaceBuyersBlock` | Gets list of users who purchased a ticket to the requested Space | I do not have a monetized account for this | ✅
`TwitterGetSpaceTweetsBlock` | Gets list of Tweets shared in the requested Space | No error | ✅
`TwitterUnfollowListBlock` | Unfollows a Twitter list for the authenticated user | No error | ✅
`TwitterFollowListBlock` | Follows a Twitter list for the authenticated user | No error | ✅
`TwitterListGetFollowersBlock` | Gets followers of a specified Twitter list | Enterprise level access | ❌
`TwitterGetFollowedListsBlock` | Gets lists followed by a specified Twitter user | Enterprise level access | ❌
`TwitterGetListBlock` | Gets information about a Twitter List specified by ID | No error | ✅
`TwitterGetOwnedListsBlock` | Gets all Lists owned by the specified user | No error | ✅
`TwitterRemoveListMemberBlock` | Removes a member from a Twitter List that the authenticated user owns | No error | ✅
`TwitterAddListMemberBlock` | Adds a member to a Twitter List that the authenticated user owns | No error | ✅
`TwitterGetListMembersBlock` | Gets the members of a specified Twitter List | No error | ✅
`TwitterGetListMembershipsBlock` | Gets all Lists that a specified user is a member of | No error | ✅
`TwitterGetListTweetsBlock` | Gets tweets from a specified Twitter list | No error | ✅
`TwitterDeleteListBlock` | Deletes a Twitter List owned by the authenticated user | No error | ✅
`TwitterUpdateListBlock` | Updates a Twitter List owned by the authenticated user | No error | ✅
`TwitterCreateListBlock` | Creates a Twitter List owned by the authenticated user | No error | ✅
`TwitterUnpinListBlock` | Enables the authenticated user to unpin a List. | No error | ✅
`TwitterPinListBlock` | Enables the authenticated user to pin a List. | No error | ✅
`TwitterGetPinnedListsBlock` | Returns the Lists pinned by the authenticated user. | No error | ✅
`TwitterGetDMEventsBlock` | Gets a list of Direct Message events for the authenticated user | Need Enterprise level access | ❌
`TwitterSendDirectMessageBlock` | Sends a direct message to a Twitter user | Need Enterprise level access | ❌
`TwitterCreateDMConversationBlock` | Creates a new group direct message | Need Enterprise level access | ❌

### Need to add more stuff
1. A normal input to select date and time.
2. Some more enterprise-level blocks, especially webhook triggers.

Supported triggers 


Event Name | Description
-- | --
Posts (by user) | User creates a new post.
Post deletes (by user) | User deletes an existing post.
@mentions (of user) | User is mentioned in a post.
Replies (to or from user) | User replies to a post or receives a reply from another user.
Retweets (by user or of user) | User retweets a post or someone retweets the user's post.
Quote Tweets (by user or of user) | User quote tweets a post or someone quote tweets the user's post.
Retweets of Quoted Tweets (by user or of user) | Retweets of quote tweets by the user or of the user.
Likes (by user or of user) | User likes a post or someone likes the user's post.
Follows (by user or of user) | User follows another user or another user follows the user.
Unfollows (by user) | User unfollows another user.
Blocks (by user) | User blocks another user.
Unblocks (by user) | User unblocks a previously blocked user.
Mutes (by user) | User mutes another user.
Unmutes (by user) | User unmutes a previously muted user.
Direct Messages sent (by user) | User sends direct messages to other users.
Direct Messages received (by user) | User receives direct messages from other users.
Typing indicators (to user) | Indicators showing when someone is typing a message to the user.
Read receipts (to user) | Indicators showing when the user has read a message.
Subscription revokes (by user) | User revokes a subscription to a service or content.

aarushik93 commented:
wooo thanks for this. Testing & reviewing!

one thing to fix the CI, please run what it's asking for

poetry lock --no-update

Abhi1992002 commented:
Sorry, I haven't fixed some block tests after making changes. I will make the tests work by tomorrow.

Abhi1992002 commented:
@aarushik93 @Bentlybro @Torantulino 

The main issue in this file is with the node-input-component file.
I have added a multi-select input for a list of enums, and @Pwuts has also made a change to that.
Could you please review my changes in the select file to see if any adjustments are needed? If there are, I will make the changes.

**Summary**:

- I have added an extra option in the input `is_multi_select` and `enum`.
- If the input type is an array, is multi-selected, and has an enum, I am rendering the multi-select.

Pwuts commented on autogpt_platform/autogpt_libs/autogpt_libs/supabase_integration_credentials_store/store.py:
```diff
@@ -263,6 +288,26 @@ def get_any_valid_scopes_from_state_token(
 
         return []
 
+    def _get_code_verifier(self, user_id: str, provider: str, token : str) -> Optional[str]:
+        user_integrations = self._get_user_integrations(user_id)
+        oauth_states = user_integrations.oauth_states
+
+        now = datetime.now(timezone.utc)
+        valid_state = next(
+            (
+                state
+                for state in oauth_states
+                if state.token == token
+                and state.provider == provider
+                and state.expires_at > now.timestamp()
+            ),
+            None,
+        )
+
+        if valid_state:
+            return valid_state.code_verifier
```
Please combine this with `get_any_valid_scopes_from_state_token`

Pwuts commented on autogpt_platform/autogpt_libs/autogpt_libs/supabase_integration_credentials_store/store.py:
```diff
@@ -210,18 +212,24 @@ def delete_creds_by_id(self, user_id: str, credentials_id: str) -> None:
         token = secrets.token_urlsafe(32)
         expires_at = datetime.now(timezone.utc) + timedelta(minutes=10)
 
+        code_verifier = self._generate_code_verifier(128)
+        code_challenge = self._generate_code_challenge(code_verifier)
```
These can be combined into one method

Pwuts commented on autogpt_platform/backend/backend/data/model.py:
```diff
@@ -112,6 +112,7 @@ def SchemaField(
     placeholder: Optional[str] = None,
     advanced: Optional[bool] = None,
     secret: bool = False,
+    is_multi_select: bool = False,
```
This seems unnecessary to me. `list[SomeEnum]` is always going to be multi-select.

Pwuts commented on autogpt_platform/backend/backend/server/integrations/router.py:
```diff
@@ -69,6 +70,9 @@ def callback(
 ) -> CredentialsMetaResponse:
     logger.debug(f"Received OAuth callback for provider: {provider}")
     handler = _get_provider_oauth_handler(request, provider)
+    code_verifier = creds_manager.store._get_code_verifier(
+        user_id, provider, state_token
+    )
```
1. Please do this after the state token is verified
2. The leading underscore suggests this method is only for internal use, but here it's called externally

aarushik93 commented on autogpt_platform/backend/backend/blocks/twitter/direct_message/direct_message_lookup.py:
```diff
@@ -0,0 +1,199 @@
+# from typing import cast
+# import tweepy
```
why are these files commented out?

Abhi1992002 replied:
Since they are not manually tested, they might have some bugs. We need Enterprise/professional-level access for DMs.

Pwuts commented on autogpt_platform/backend/backend/server/integrations/router.py:
```diff
@@ -69,6 +70,9 @@ def callback(
+    )
 
     # Verify the state token
     if not creds_manager.store.verify_state_token(user_id, state_token, provider):
```
Could you please make `verify_state_token` return the `OAuthState` object and use that instead of the two separate `_get_code_verifier` and `get_any_valid_scopes_from_state_token` calls?

aarushik93 commented on autogpt_platform/frontend/src/components/CustomNode.tsx:
```diff
@@ -262,6 +262,9 @@ export function CustomNode({
     const newValues = JSON.parse(JSON.stringify(data.hardcodedValues));
     let current = newValues;
 
+    console.log("Keys: ", keys);
```
can these logs be removed?

Pwuts commented on autogpt_platform/autogpt_libs/autogpt_libs/supabase_integration_credentials_store/store.py:
```diff
@@ -210,18 +212,24 @@ def delete_creds_by_id(self, user_id: str, credentials_id: str) -> None:
         state = OAuthState(
             token=token,
             provider=provider,
+            code_verifier= code_verifier,
```
I'd prefer if we can somehow omit this if the provider doesn't support it

Pwuts replied:
It's fine, if omitting it adds significant complexity just leave it.

aarushik93 commented on autogpt_platform/backend/.env.example:
```diff
@@ -51,6 +51,11 @@ GITHUB_CLIENT_SECRET=
 GOOGLE_CLIENT_SECRET=
 
+# Twitter/X OAuth App server credentials - https://developer.x.com/en/products/x-api
+Twitter_CLIENT_ID=
```
Is this supposed to be all caps?

Abhi1992002 replied:
sorry for this silly one

aarushik93 commented:
@Abhi1992002 im testing this PR and not sure where or what user id is? is it username or different?

Abhi1992002 commented on autogpt_platform/backend/backend/integrations/oauth/twitter.py:
```diff
@@ -0,0 +1,171 @@
+        # scopes = self.handle_default_scopes(scopes)
+
+        if code_challenge is None:
+            raise ValueError("code_verifier is required for Twitter OAuth")
```
Done..!!

Abhi1992002 commented:
TODO: [will do it tomorrow] 

- [ ] Use X instead of Twitter
- [ ] Fix multi select support.

Abhi1992002 commented:
TODO:
- [ ] Add `depend_on` and `oneof` field support 
- [ ] handle Optional values in a better way on frontend
- [ ] All fields are required right now, Need to make some Optional
- [ ] Debug and test some Enterprise Level Blocks

ntindle commented on autogpt_platform/backend/.env.example:
```diff
@@ -58,6 +58,11 @@ GITHUB_CLIENT_SECRET=
 GOOGLE_CLIENT_SECRET=
 
+# Twitter/X OAuth App server credentials - https://developer.x.com/en/products/x-api
+TWITTER_CLIENT_ID=
```
Be more clearr with what these are and where to get them. Do I need to setup User Authentication Settings?

ntindle commented:
![image](https://github.com/user-attachments/assets/cfcbbd6b-c429-41ae-9ea5-8a8dd17bbfa9)
These two filters need much better descriptions and are also required for some reason even though the text of the expansions is off

ntindle commented:
fixed and pushed
- Credential fixes
- Linting fixes
- Poetry fixes (stuff was updated that probably shouldn't have been)

ntindle commented on autogpt_platform/backend/backend/blocks/twitter/_auth.py:
```diff
@@ -0,0 +1,60 @@
+    """
+    return CredentialsField(
+        # required_scopes=set(scopes),
+        required_scopes=set(TwitterOAuthHandler.DEFAULT_SCOPES + scopes),
```
pointing out that I added `+ scopes` here. not sure if it was needed but it was just taking the set of the default scopes which is weird
